src/EditorFeatures/Core/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,180 @@
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.Editor.Commands;
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.SuggestionSupport;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Outlining;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    [ExportCommandHandler("Go to next and previous method command handler", ContentTypeNames.RoslynContentType)]
+    class GoToNextAndPreviousMethodCommandHandler : ICommandHandler<GoToNextMethodCommandArgs>, ICommandHandler<GoToPreviousMethodCommandArgs>
+    {
+        private readonly IWaitIndicator _waitIndicator;
+        private readonly IOutliningManagerService _outliningManagerService;
+
+        [ImportingConstructor]
+        public GoToNextAndPreviousMethodCommandHandler(IWaitIndicator waitIndicator, IOutliningManagerService outliningManagerService)
+        {
+            _waitIndicator = waitIndicator;
+            _outliningManagerService = outliningManagerService;
+        }
+
+        public CommandState GetCommandState(GoToNextMethodCommandArgs args, Func<CommandState> nextHandler)
+            => GetCommandState(args.SubjectBuffer, nextHandler);
+
+        public void ExecuteCommand(GoToNextMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: true);
+
+        public CommandState GetCommandState(GoToPreviousMethodCommandArgs args, Func<CommandState> nextHandler)
+            => GetCommandState(args.SubjectBuffer, nextHandler);
+
+        public void ExecuteCommand(GoToPreviousMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: false);
+
+        private static CommandState GetCommandState(ITextBuffer subjectBuffer, Func<CommandState> nextHandler)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var documentSupportsSuggestionService = document.Project.Solution.Workspace.Services.GetService<IDocumentSupportsSuggestionService>();
+            if (document == null || !document.SupportsSyntaxTree || !documentSupportsSuggestionService.SupportsRefactorings(document))
+            {
+                return nextHandler();
+            }
+
+            return CommandState.Available;
+        }
+
+
+        private void ExecuteCommand(Action nextHandler, ITextBuffer subjectBuffer, ITextView textView, bool next)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
Check ```SupportsGoToNextPreviousMethod``` (and all the things checked in GetCommandState) again here. Wouldn't want to have passed through the GetCommandState to nextHandler due to some failed check but then go ahead and attempt to handle the actual command here anyway.